Add T5 Encoder for Feature Extraction

agemagician opened the pull request:
# What does this PR do?

While using T5 for feature extraction, I found out that T5 encoder provides better features than T5 decoder. Hence, it makes sense to have T5 encoder only, which should reduce the memory and inference time by half, if feature extraction is needed rather than conditional generation. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

 T5: @patrickvonplaten
 tensorflow: @jplu

patrickvonplaten commented:
I like it!

patrickvonplaten commented on src/transformers/models/t5/modeling_t5.py:
```diff
@@ -1329,3 +1329,93 @@ def _reorder_cache(self, past, beam_idx):
+            self.encoder.layer[layer].attention.prune_heads(heads)
+
+    @add_start_docstrings_to_model_forward(T5_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=Seq2SeqModelOutput, config_class=_CONFIG_FOR_DOC)
```
```suggestion
    @replace_return_docstrings(output_type=BaseModelOutput, config_class=_CONFIG_FOR_DOC)
```

patrickvonplaten commented on src/transformers/models/t5/modeling_t5.py:
```diff
@@ -1329,3 +1329,93 @@ def _reorder_cache(self, past, beam_idx):
+        for layer, heads in heads_to_prune.items():
+            self.encoder.layer[layer].attention.prune_heads(heads)
+
+    @add_start_docstrings_to_model_forward(T5_INPUTS_DOCSTRING)
```
I think `T5_INPUTS_DOCSTRING` is includes all the decoder inputs as well -> I think we'll have to make a new `T5_ENCODER_INPUTS_DOCSTRING` here that only includes the Encoder inputs.

agemagician replied:
done. I have add it.

patrickvonplaten commented on src/transformers/models/t5/modeling_t5.py:
```diff
@@ -1329,3 +1329,93 @@ def _reorder_cache(self, past, beam_idx):
+class T5ModelEncoder(T5PreTrainedModel):
+    authorized_missing_keys = [
+        r"encoder\.embed_tokens\.weight",
+        r"decoder\.embed_tokens\.weight",
```
no need for `decoder`

agemagician replied:
done for all decoder keys.
Let me know if you have a better regex.

patrickvonplaten commented on src/transformers/models/t5/modeling_t5.py:
```diff
@@ -1329,3 +1329,93 @@ def _reorder_cache(self, past, beam_idx):
+        self,
+        input_ids=None,
+        attention_mask=None,
+        encoder_outputs=None,
```
`encoder_outputs` are only for Seq2Seq I believe

agemagician replied:
done. I removed it.

patrickvonplaten commented on src/transformers/models/t5/modeling_t5.py:
```diff
@@ -1329,3 +1329,93 @@ def _reorder_cache(self, past, beam_idx):
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        # Encode if needed (training, first prediction pass)
+        if encoder_outputs is None:
```
we don't need this `if statement` => `encoder_outputs should always be `None` otherwise the forward pass is useless

agemagician replied:
done. I removed it.

patrickvonplaten commented on src/transformers/models/t5/modeling_t5.py:
```diff
@@ -1329,3 +1329,93 @@ def _reorder_cache(self, past, beam_idx):
+                output_hidden_states=output_hidden_states,
+                return_dict=return_dict,
+            )
+        elif return_dict and not isinstance(encoder_outputs, BaseModelOutput):
```
```suggestion
        if return_dict:
```

agemagician replied:
done. I removed it.

patrickvonplaten commented on src/transformers/models/t5/modeling_t5.py:
```diff
@@ -1329,3 +1329,93 @@ def _reorder_cache(self, past, beam_idx):
+            )
+        elif return_dict and not isinstance(encoder_outputs, BaseModelOutput):
+            encoder_outputs = BaseModelOutput(
+                last_hidden_state=encoder_outputs[0],
```
Here we can change to `encoder_outputs.last_hidden_state`, etc....

agemagician replied:
I am not sure, but I think you meant to change encoder_outputs[0] to encoder_outputs.last_hidden_state.
I have updated it.

patrickvonplaten commented on src/transformers/models/t5/modeling_tf_t5.py:
```diff
@@ -1410,3 +1410,124 @@ def _reorder_cache(self, past, beam_idx) -> Tuple:
 
             reordered_decoder_past = reordered_decoder_past + (reordered_layer_past_states,)
         return past + (reordered_decoder_past,)
+
```
same changes should be done for TF

agemagician replied:
Done. Hopefully, I didn't miss anything.

patrickvonplaten reviewed (commented):
I like this addition a lot! 

We should also include this new model in the documentation here:
https://github.com/huggingface/transformers/blob/master/docs/source/model_doc/t5.rst#t5model
and add it to the tests here:
https://github.com/huggingface/transformers/blob/2c83b3c38d00377c0bb92a7f438758e10417bfd1/tests/test_modeling_t5.py#L474

=> There might be some issues with the tests after adding the model...I'm happy to go into your PR and fix them accordingly :-) 

And we can then also add the model to `mT5` :-)

jplu reviewed (commented):
Thank you very much for this addition!!

@patrickvonplaten is the T5 master here so I let him review for the modeling point of view. About TF I think we should wait that the PR about the new inputs to be merged. Also a tiny comment.

agemagician commented on src/transformers/models/t5/modeling_tf_t5.py:
```diff
@@ -1410,3 +1410,124 @@ def _reorder_cache(self, past, beam_idx) -> Tuple:
+        output_hidden_states=None,
+        return_dict=None,
+        training=False,
+        **kwargs,
```
Great, I am glad that you did like it.

Thanks @patrickvonplaten and @jplu for your feedback.

@patrickvonplaten :
I have adjusted all your code review, and also add it to the T5 documentation.
The only missing part is the tests, it will be great if you could add it.

@jplu :
I have removed them from TF model.

Is there anything else needed from my side to merge the pull request ?

agemagician commented:
Great, I am glad that you did like it.

Thanks @patrickvonplaten and @jplu for your feedback.

@patrickvonplaten :
I have adjusted all your code review, and also add it to the T5 documentation.
The only missing part is the tests, it will be great if you could add it.

@jplu :
I have removed them from TF model.

Is there anything else needed from my side to merge the pull request ?

patrickvonplaten commented:
> Great, I am glad that you did like it.
> 
> Thanks @patrickvonplaten and @jplu for your feedback.
> 
> @patrickvonplaten :
> I have adjusted all your code review, and also add it to the T5 documentation.
> The only missing part is the tests, it will be great if you could add it.
> 
> @jplu :
> I have removed the unnecessary parameters from the TF model.
> 
> Is there anything else needed from my side to merge the pull request ?

I think that's great! I'll fix the tests and merge :-)

patrickvonplaten commented on tests/test_modeling_rag.py:
```diff
@@ -554,7 +554,7 @@ class RagDPRT5Test(RagTestMixin, unittest.TestCase):
     def config_and_inputs(self):
         question_encoder_tester = DPRModelTester(self)
         dpr_config_and_inputs = question_encoder_tester.prepare_config_and_inputs()
-        generator_tester = T5ModelTester(self, vocab_size=1100, n_positions=30)
```
`n_positions` does not exist anymore => it was useless so remove it from tests as well

patrickvonplaten commented:
## Update:

PR is ready for review IMO. Would be great if @LysandreJik @jplu and @sgugger you can take a look :-)

agemagician commented:
> ## Update:
> PR is ready for review IMO. Would be great if @LysandreJik @jplu and @sgugger you can take a look :-)

Thanks a lot @patrickvonplaten ^_^

LysandreJik commented on src/transformers/models/mt5/modeling_mt5.py:
```diff
@@ -73,11 +73,32 @@ class MT5ForConditionalGeneration(T5ForConditionalGeneration):
+    This class overrides :class:`~transformers.T5EncoderModel`. Please check the superclass for the appropriate
+    documentation alongside usage examples.
+
+    Examples::
```
```suggestion
    Examples::
    
```
There needs to be two line returns between the `Examples::` and the actual example.

LysandreJik commented on src/transformers/models/mt5/modeling_tf_mt5.py:
```diff
@@ -64,3 +64,22 @@ class TFMT5ForConditionalGeneration(TFT5ForConditionalGeneration):
+    This class overrides :class:`~transformers.TFT5EncoderModel`. Please check the superclass for the appropriate
+    documentation alongside usage examples.
+
+    Examples::
```
Same here

LysandreJik replied:
```suggestion
    Examples::
    
```

LysandreJik commented on src/transformers/models/t5/modeling_t5.py:
```diff
@@ -1518,3 +1557,79 @@ def _reorder_cache(self, past, beam_idx):
+        r"""
+        Returns:
+
+        Example::
```
```suggestion
        Example::
        
```

LysandreJik commented on src/transformers/models/t5/modeling_tf_t5.py:
```diff
@@ -949,6 +949,48 @@ def _shift_right(self, input_ids):
+            Indices of input sequence tokens in the vocabulary. T5 is a model with relative position embeddings so you
+            should be able to pad the inputs on the right or the left.
+
+            Indices can be obtained using :class:`~transformers.BertTokenizer`. See
```
Should be the T5 tokenizer
```suggestion
            Indices can be obtained using :class:`~transformers.T5Tokenizer`. See
```

patrickvonplaten replied:
Great catch!

LysandreJik reviewed (approved):
Very clean implementation, thanks a lot @agemagician!

jplu commented on tests/test_modeling_tf_t5.py:
```diff
@@ -295,6 +295,128 @@ def test_saved_model_with_hidden_states_output(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_model(*config_and_inputs)
+
+    # is not able to be part of a pipeline
```
What does-it means _not able to be part of a pipeline_? It is because the test fails?

patrickvonplaten replied:
Since T5Encoder just outputs the encoded hidden states I cannot really be used in combination with `pipeline()`

jplu reviewed (approved):
LGTM! Great work!!

agemagician commented:
> Very clean implementation, thanks a lot @agemagician!

You are welcome.
I am glad that I could help making the library better, even with a small contribution.
